### PR TITLE
fix: Track referrer/search params per browser session

### PR DESCRIPTION
**Pull request by mariusandra**
## Changes

- Resolves https://github.com/PostHog/posthog-js/issues/477
- Adds `sessionRegister` and `sessionPersistence`, which persists properties only until the user closes their browser.
- Uses this to persist UTM properties and other session-related data. These would normally be sent with every event, even months later.

## Checklist
- [ ] Tests for new code (see [advice on the tests we use](https://github.com/PostHog/posthog-js#tiers-of-testing))
- [ ] Accounted for the impact of any changes across different browsers

### CODE REVIEW

**Comment by mariusandra** [comment]
I got it far enough it could get a look. Tagged @pauldambra because help!, and @benjackwhite because sessions. 

TODO and next steps. All help appreciated, especially regarding the docs and communication.

- Communicate the change to users, as people might be relying on the old behaviour, and will be surprised if this changes randomly.
- Decide: Should we provide a fallback to revert to the old behaviour? E.g. a config option or setting?
- Decide: Should we _yeet_ these campaign properties from people's normal super properties as well?
- Update the integration docs and add a section about session properties.

File: src/posthog-core.ts [hunk 1]
```diff
@@ -169,14 +169,14 @@ const defaultConfig = (): PostHogConfig => ({
 })
 
 /**
- * create_mplib(token:string, config:object, name:string)
+ * create_phlib(token:string, config:object, name:string)
```
Comment (pauldambra) [inline src/posthog-core.ts:172]: 👨‍🍳 👌

File: src/posthog-core.ts [hunk 2]
```diff
@@ -1113,6 +1168,7 @@ export class PostHog {
     reset(reset_device_id?: boolean): void {
```
Comment (pauldambra) [inline src/posthog-core.ts:1168]: Should we reset this when the session id resets?

**Review by pauldambra — COMMENTED** [review]
Quick pass over, looks good. Not tested the running code yet

File: src/posthog-core.ts (hunk 2 above)
Comment (benjackwhite) [reply src/posthog-core.ts:1168]: I don't think we should. The SessionStorage is basically "as long as the window is open" which is not all that long (unless you are @neilkakkar...)

**Review by benjackwhite — APPROVED** [review]
Looks good to me.

File: src/posthog-core.ts (hunk 1 above)
Comment (Twixes) [reply src/posthog-core.ts:174]: I applaud the change but it seems quite irrelevant to this PR per se, and has caused some unnecessary merge conflicts. Small independent `refactor:` PRs for the win, in the future ;)

**Comment by Twixes** [comment]
Pulled in support for `sessionStorage` in the opt-in-out classes (from #592), improved typing, and clarified some names (to e.g. `registerSession` is now `register_for_session` to align it with `register_once` and reduce conflict with the PostHog concept of a session).
Take a look if all of this looks good @mariusandra!

File: src/types.ts [hunk 3]
```diff
@@ -54,7 +54,7 @@ export interface PostHogConfig {
     autocapture: boolean | AutocaptureConfig
     rageclick: boolean
     cross_subdomain_cookie: boolean
-    persistence: 'localStorage' | 'cookie' | 'memory' | 'localStorage+cookie'
+    persistence: 'localStorage' | 'cookie' | 'memory' | 'localStorage+cookie' | 'sessionStorage'
```
Comment (mariusandra) [inline src/types.ts:57]: Might be worth aliasing this long and twice repeated `|` list as `type PersistenceType` or such?

**Review by mariusandra — COMMENTED** [review]
Looks good to me! Can't ✅ my own PR though.

The lowercasing part with `persistence` seems a bit yucky though, but I can't suggest anything much better, so good enough :).

File: src/types.ts (hunk 3 above)
Comment (Twixes) [reply src/types.ts:57]: It's only repeated once as far as I can see, can you give me a pointer?

Comment (Twixes) [reply src/types.ts:57]: Ah, I suppose the GDRP/OptInOut config? On second thought, it's probably _not_ a good idea to allow that, as those settings SHOULD be persisted. Narrowed that down to just `'cookie' | 'localStorage'| 'localStorage+cookie'`.

**Pull request merged**